[SPIRV] Add FPEncoding operand support for OpTypeFloat

YixingZhang007 opened the pull request:
This PR introduces support for `FPEncoding` operand for SPIR-V instruction `OpTypeFloat`, with the following main changes:
1. Introduces `FPEncoding` enum class to represent floating-point encodings, such as `BFloat16KHR`, in SPIR-V.
2. Updates SPIR-V instruction `OpTypeFloat` to accept `FPEncoding` as its second input operand.
3. Updates SPIR-V type creation to handle new encoding requirements.

This PR enables support for the BFloat floating-point type in SPIR-V.

YixingZhang007 committed:
add the support for bfloat in SPIRV

maarquitos14 reviewed (approved):
A couple of nits, but looks good to me otherwise.

YixingZhang007 commented on llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.cpp:
```diff
@@ -2088,3 +2100,15 @@ bool SPIRVGlobalRegistry::hasBlockDecoration(SPIRVType *Type) const {
+SPIRVGlobalRegistry::FPVariant
+SPIRVGlobalRegistry::getFPVariantForVReg(Register VReg,
+                                         const MachineFunction *MF) {
+  auto t = VRegFPVariantMap.find(MF ? MF : CurMF);
```
Thanks for the suggestion! I have made the change :)

maarquitos14 commented on llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.cpp:
```diff
@@ -2088,3 +2100,21 @@ bool SPIRVGlobalRegistry::hasBlockDecoration(SPIRVType *Type) const {
+                                         const MachineFunction *MF) {
+  const MachineFunction *Func = MF ? MF : CurMF;
+  DenseMap<const MachineFunction *,
+           DenseMap<Register, FPVariant>>::const_iterator FuncIt =
```
I didn't realize it was such a nasty type. You can go back to auto if you wish.

s-perron reviewed (commented):
I don't have any problems with this.

Keenuts reviewed (commented):
No issues with this, but seconding the question about the reuse of the existing `VRegToTypeMap`. Is there a performance need to store the FPVariant in a map, or can getFPVariantForVreg do a lookup in the generic map and then do the `isBFloatTy`/`isFloatTy` every time?

YixingZhang007 commented:
I’ve updated this PR by changing the approach for tracking the type of floating-point registers (IEEE 754 vs. Bfloat).
Previously, we used an `FPVariant` class along with a `VRegFPVariantMap` to map registers to their floating-point type. Now, I’ve replaced these with an `FPEncoding` enum class and modified the SPIR-V instruction format for floating point registers from `OpTypeFloat $bitwidth` to `OpTypeFloat $bitwidth $FPEncoding`. This way, the floating-point type can be directly distinguished from the second operand of the OpTypeFloat instruction.

YixingZhang007 commented on llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.cpp:
```diff
@@ -2088,3 +2100,17 @@ bool SPIRVGlobalRegistry::hasBlockDecoration(SPIRVType *Type) const {
+    if (VRegIt != VRegMap.end())
+      return VRegIt->second;
+  }
+  return FPVariant::NONE;
```
Thanks for the suggestion! I’ve updated the approach for representing floating-point types in the PR, and `FPVariant` is no longer used. Please let me know if there’s anything I can improve with the current approach. Thank you! :)

YixingZhang007 commented on llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.h:
```diff
@@ -88,6 +92,11 @@ class SPIRVGlobalRegistry : public SPIRVIRMapping {
 
+  // Maps floating point Registers to their FPVariant (float type kind), given
+  // the MachineFunction.
+  DenseMap<const MachineFunction *, DenseMap<Register, FPVariant>>
```
Sorry that there was a mistake in my initial thinking. I thought `OpTypeFloat` was a non-variadic SPIR-V operation, so all floating points would be represented with instruction `OpTypeFloat 16`. I have updated `OpTypeFloat` to now be a variadic operation. Thanks  :)

maarquitos14 reviewed (approved):
LGTM. Should we add any test to see if we can properly generate bfloat16 type?

Keenuts reviewed (approved):
Thanks!

MrSidims reviewed (approved):
Test failure is unrelated and will be handled separately.
